File: JavaScript/Medium/PalindromicSubstrings.js
/**
 * @param {string} s
 * @return {number}
 */
var countSubstrings = function(s) {
    
  let count = 0;
  
  for (let i = 0; i < s.length; i++) {
    count += expandFromCenter(s, i, i);
    count += expandFromCenter(s, i, i + 1);
  }
  
  return count;
};

let expandFromCenter = function(s, left, right) {
  let count = 0;
  while (left >= 0 && right < s.length && s.charAt(left) === s.charAt(right)) { 
    left -= 1;
    right += 1;
    count += 1;
  }
  return count;
}

/*

    Expand from the center of each index, the center could be between two letters.

    Time: O(n^2)
    Space: O(1)

*/

File: C#/Medium/PalindromicSubstrings.cs
public class Solution
{
    public int CountSubstrings(String s)
    {
        int count = 0;
        for (int i = 0; i < s.Length; i++)
        {
            count += ExpandFromCenter(i, i, s) + ExpandFromCenter(i, i + 1, s);
        }

        return count;
    }

    private int ExpandFromCenter(int left, int right, string s)
    {
        int count = 0;
        while (left >= 0 && right < s.Length && s[left] == s[right])
        {
            count++;
            left--;
            right++;
        }

        return count;
    }
}

/*

    Expand from the center of each index, the center could be between two letters.

    Time: O(n^2)
    Space: O(1)

*/
